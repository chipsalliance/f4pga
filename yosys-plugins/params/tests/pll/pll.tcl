yosys -import
if { [info procs getparam] == {} } { plugin -i params }
yosys -import  ;# ingest plugin commands

read_verilog $::env(DESIGN_TOP).v
read_verilog -specify -lib -D_EXPLICIT_CARRY +/xilinx/cells_sim.v
read_verilog -lib +/xilinx/cells_xtra.v
hierarchy -check -auto-top

# Check phase parameter values on bith PLLE2_ADV instances
set reference_phase [list 90 70]
set phase [getparam CLKOUT2_PHASE top/PLLE2_ADV_0 top/PLLE2_ADV]
if {[llength $phase] != 2} {
	error "Getparam should return a list with 2 elements"
}
set fp [open [test_output_path "pll.txt"] "w"]
puts -nonewline $fp "Phase before: "
if {$phase == $reference_phase} {
	puts $fp "PASS"
} else {
	puts $fp "FAIL: $phase != $reference_phase"
}

# Modify the phase parameter value on one of the PLLE2_ADV instances
setparam -set CLKOUT2_PHASE [expr [lindex $phase 0] * 1000] top/PLLE2_ADV

# Verify that the parameter has been correctly updated on the chosen instance
set reference_phase [list 90000 70]
set phase [getparam CLKOUT2_PHASE top/PLLE2_ADV_0 top/PLLE2_ADV]
puts -nonewline $fp "Phase after: "
if {$phase == $reference_phase} {
	puts $fp "PASS"
} else {
	puts $fp "FAIL: $phase != $reference_phase"
}
close $fp

# Start flow after library reading
synth_xilinx -flatten -abc9 -nosrl -noclkbuf -nodsp -iopad -run prepare:check

# Map Xilinx tech library to 7-series VPR tech library.
read_verilog -lib [file dirname $::env(DESIGN_TOP)]/techmaps/cells_sim.v
techmap -map  [file dirname $::env(DESIGN_TOP)]/techmaps/cells_map.v

# opt_expr -undriven makes sure all nets are driven, if only by the $undef
# net.
opt_expr -undriven
opt_clean

setundef -zero -params
stat

# Clean processes before writing JSON.
yosys proc

# Write the design in JSON format.
write_json [test_output_path "pll.json"]
write_blif -attr -param -cname -conn [test_output_path "pll.eblif"]
