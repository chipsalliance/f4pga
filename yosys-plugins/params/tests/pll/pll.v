// Copyright 2020-2022 F4PGA Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

module top (
    (* dont_touch = "true" *) input clk100,
    input cpu_reset,
    output [2:0] led
);

  wire [2:0] led;
  wire builder_pll_fb;

  assign led[0] = main_locked;
  assign led[1] = main_clkout0;
  assign led[2] = main_clkout1;

  PLLE2_ADV #(
      .CLKFBOUT_MULT(4'd12),
      .CLKIN1_PERIOD(10.0),
      .CLKOUT0_DIVIDE(5'd20),
      .CLKOUT0_PHASE(1'd0),
      .CLKOUT1_DIVIDE(3'd5),
      .CLKOUT1_PHASE(1'd0),
      .CLKOUT2_DIVIDE(3'd5),
      .CLKOUT2_PHASE(7'd70),
      .CLKOUT3_DIVIDE(3'd6),
      .CLKOUT3_PHASE(1'd0),
      .DIVCLK_DIVIDE(1'd1),
      .REF_JITTER1(0.01),
      .STARTUP_WAIT("FALSE")
  ) PLLE2_ADV_0 (
      .CLKFBIN(builder_pll_fb),
      .CLKIN1(clk100),
      .RST(cpu_reset),
      .CLKFBOUT(builder_pll_fb),
      .CLKOUT0(main_clkout0),
      .CLKOUT1(main_clkout1),
      .LOCKED(main_locked)
  );
  PLLE2_ADV #(
      .CLKFBOUT_MULT(4'd12),
      .CLKIN1_PERIOD(10.0),
      .CLKOUT0_DIVIDE(5'd20),
      .CLKOUT0_PHASE(1'd0),
      .CLKOUT1_DIVIDE(3'd5),
      .CLKOUT1_PHASE(1'd0),
      .CLKOUT2_DIVIDE(3'd5),
      .CLKOUT2_PHASE(7'd90),
      .CLKOUT3_DIVIDE(3'd6),
      .CLKOUT3_PHASE(1'd0),
      .DIVCLK_DIVIDE(1'd1),
      .REF_JITTER1(0.01),
      .STARTUP_WAIT("FALSE")
  ) PLLE2_ADV (
      .CLKFBIN(builder_pll_fb),
      .CLKIN1(clk100),
      .RST(cpu_reset),
      .CLKFBOUT(builder_pll_fb),
      .CLKOUT0(main_clkout0),
      .CLKOUT1(main_clkout1),
      .LOCKED(main_locked)
  );

endmodule
