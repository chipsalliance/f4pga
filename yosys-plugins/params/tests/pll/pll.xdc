# ## clk100:0
set_property LOC E3 [get_ports clk100]
set_property IOSTANDARD LVCMOS33 [get_ports clk100]
# ## cpu_reset:0
set_property LOC C2 [get_ports cpu_reset]
set_property IOSTANDARD LVCMOS33 [get_ports cpu_reset]
set_property LOC H5 [get_ports {led[0]}]
set_property LOC J5 [get_ports {led[1]}]
set_property LOC T9 [get_ports {led[2]}]
