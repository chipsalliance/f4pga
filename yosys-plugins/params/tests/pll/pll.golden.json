{
  "PLLE2_ADV": {
    "CLKFBOUT_CLKOUT1_HIGH_TIME": "000110",
    "CLKFBOUT_CLKOUT1_LOW_TIME": "000110",
    "CLKFBOUT_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKFBOUT_CLKOUT1_PHASE_MUX": "000",
    "CLKFBOUT_CLKOUT2_DELAY_TIME": "000000",
    "CLKFBOUT_CLKOUT2_EDGE": "0",
    "CLKFBOUT_CLKOUT2_NO_COUNT": "0",
    "CLKOUT0_CLKOUT1_HIGH_TIME": "001010",
    "CLKOUT0_CLKOUT1_LOW_TIME": "001010",
    "CLKOUT0_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKOUT0_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT0_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT0_CLKOUT2_EDGE": "0",
    "CLKOUT0_CLKOUT2_NO_COUNT": "0",
    "CLKOUT1_CLKOUT1_HIGH_TIME": "000010",
    "CLKOUT1_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT1_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKOUT1_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT1_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT1_CLKOUT2_EDGE": "1",
    "CLKOUT1_CLKOUT2_NO_COUNT": "0",
    "CLKOUT2_CLKOUT1_HIGH_TIME": "000010",
    "CLKOUT2_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT2_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT2_CLKOUT1_PHASE_MUX": "010",
    "CLKOUT2_CLKOUT2_DELAY_TIME": "000001",
    "CLKOUT2_CLKOUT2_EDGE": "1",
    "CLKOUT2_CLKOUT2_NO_COUNT": "0",
    "CLKOUT3_CLKOUT1_HIGH_TIME": "000011",
    "CLKOUT3_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT3_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT3_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT3_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT3_CLKOUT2_EDGE": "0",
    "CLKOUT3_CLKOUT2_NO_COUNT": "0",
    "CLKOUT4_CLKOUT1_HIGH_TIME": "000001",
    "CLKOUT4_CLKOUT1_LOW_TIME": "000001",
    "CLKOUT4_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT4_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT4_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT4_CLKOUT2_EDGE": "0",
    "CLKOUT4_CLKOUT2_NO_COUNT": "1",
    "CLKOUT5_CLKOUT1_HIGH_TIME": "000001",
    "CLKOUT5_CLKOUT1_LOW_TIME": "000001",
    "CLKOUT5_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT5_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT5_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT5_CLKOUT2_EDGE": "0",
    "CLKOUT5_CLKOUT2_NO_COUNT": "1",
    "DIVCLK_DIVCLK_EDGE": "0",
    "DIVCLK_DIVCLK_HIGH_TIME": "000001",
    "DIVCLK_DIVCLK_LOW_TIME": "000001",
    "DIVCLK_DIVCLK_NO_COUNT": "1",
    "FILTREG1_RESERVED": "000000001000",
    "INV_CLKINSEL": "0",
    "LKTABLE": "1111111111110011100111111010010000000001",
    "LOCKREG3_RESERVED": "1",
    "STARTUP_WAIT": "0",
    "TABLE": "1111110100",
    "ZINV_PWRDWN": "1",
    "ZINV_RST": "0"
  },
  "PLLE2_ADV_0": {
    "CLKFBOUT_CLKOUT1_HIGH_TIME": "000110",
    "CLKFBOUT_CLKOUT1_LOW_TIME": "000110",
    "CLKFBOUT_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKFBOUT_CLKOUT1_PHASE_MUX": "000",
    "CLKFBOUT_CLKOUT2_DELAY_TIME": "000000",
    "CLKFBOUT_CLKOUT2_EDGE": "0",
    "CLKFBOUT_CLKOUT2_NO_COUNT": "0",
    "CLKOUT0_CLKOUT1_HIGH_TIME": "001010",
    "CLKOUT0_CLKOUT1_LOW_TIME": "001010",
    "CLKOUT0_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKOUT0_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT0_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT0_CLKOUT2_EDGE": "0",
    "CLKOUT0_CLKOUT2_NO_COUNT": "0",
    "CLKOUT1_CLKOUT1_HIGH_TIME": "000010",
    "CLKOUT1_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT1_CLKOUT1_OUTPUT_ENABLE": "1",
    "CLKOUT1_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT1_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT1_CLKOUT2_EDGE": "1",
    "CLKOUT1_CLKOUT2_NO_COUNT": "0",
    "CLKOUT2_CLKOUT1_HIGH_TIME": "000010",
    "CLKOUT2_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT2_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT2_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT2_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT2_CLKOUT2_EDGE": "1",
    "CLKOUT2_CLKOUT2_NO_COUNT": "0",
    "CLKOUT3_CLKOUT1_HIGH_TIME": "000011",
    "CLKOUT3_CLKOUT1_LOW_TIME": "000011",
    "CLKOUT3_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT3_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT3_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT3_CLKOUT2_EDGE": "0",
    "CLKOUT3_CLKOUT2_NO_COUNT": "0",
    "CLKOUT4_CLKOUT1_HIGH_TIME": "000001",
    "CLKOUT4_CLKOUT1_LOW_TIME": "000001",
    "CLKOUT4_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT4_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT4_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT4_CLKOUT2_EDGE": "0",
    "CLKOUT4_CLKOUT2_NO_COUNT": "1",
    "CLKOUT5_CLKOUT1_HIGH_TIME": "000001",
    "CLKOUT5_CLKOUT1_LOW_TIME": "000001",
    "CLKOUT5_CLKOUT1_OUTPUT_ENABLE": "0",
    "CLKOUT5_CLKOUT1_PHASE_MUX": "000",
    "CLKOUT5_CLKOUT2_DELAY_TIME": "000000",
    "CLKOUT5_CLKOUT2_EDGE": "0",
    "CLKOUT5_CLKOUT2_NO_COUNT": "1",
    "DIVCLK_DIVCLK_EDGE": "0",
    "DIVCLK_DIVCLK_HIGH_TIME": "000001",
    "DIVCLK_DIVCLK_LOW_TIME": "000001",
    "DIVCLK_DIVCLK_NO_COUNT": "1",
    "FILTREG1_RESERVED": "000000001000",
    "INV_CLKINSEL": "0",
    "LKTABLE": "1111111111110011100111111010010000000001",
    "LOCKREG3_RESERVED": "1",
    "STARTUP_WAIT": "0",
    "TABLE": "1111110100",
    "ZINV_PWRDWN": "1",
    "ZINV_RST": "0"
  }
}